File: DailyChallenge/2020/March/29.cpp
// https://leetcode.com/explore/challenge/card/march-leetcoding-challenge-2021/592/week-5-march-29th-march-31st/3689/
//
//

/**
 *  * Definition for a binary tree node.
 *   * struct TreeNode {
 *    *     int val;
 *     *     TreeNode *left;
 *      *     TreeNode *right;
 *       *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *        *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *         *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 *          * };
 *           */
class Solution {
	public:
		    vector<int> flipMatchVoyage(TreeNode* root, vector<int>& voyage) {
			            vector<int> res;
				            int i = 0;
					            return helper(root, voyage, i, res) ? res : vector<int>{-1};
						        }
		        bool helper(TreeNode* node, vector<int>& voyage, int& i, vector<int>& res) {
				        if (!node) return true;
					        if (node->val != voyage[i++]) return false;
						        if (node->left && node->left->val != voyage[i]) {
								            res.push_back(node->val);
									                return helper(node->right, voyage, i, res) && helper(node->left, voyage, i, res);
											        }
							        return helper(node->left, voyage, i, res) && helper(node->right, voyage, i, res);
								    }
};

File: DailyChallenge/2020/May/29.java
// https://leetcode.com/explore/challenge/card/may-leetcoding-challenge-2021/602/week-5-may-29th-may-31st/3760/


class Solution {
    private int result = 0;
    public int totalNQueens(int n) {
        if (n == 0 || n == 2 || n == 3) {
            return 0;
        }
        int[] cols = new int[n];
        totalNQueensHelper(n, 0, cols);
         
        return result;
    }
    private void totalNQueensHelper(int n, int row, int[] cols) {
        if (row == n) {
            result++;
            return;
        }
         
        for (int j = 0; j < n; j++) {
            if (isValid(j, row, cols)) {
                cols[row] = j; 
                totalNQueensHelper(n, row + 1, cols);
            }
        }
    }
     
    private boolean isValid(int col, int rows, int[] cols) {
        for (int i = 0; i < rows; i++) {
            if (cols[i] == col || rows - i == Math.abs(cols[i] - col)) {
                return false;
            }
        }
        return true;
    }
}
